src/CVRF/cvrf_priv.c

```diff
@@ -0,0 +1,2292 @@
+		return NULL;
+
+	ret->product_reference = NULL;
+	ret->relates_to_ref = NULL;
+	ret->full_name = cvrf_product_name_new();
+
+	return ret;
+}
+
+void cvrf_relationship_free(struct cvrf_relationship *relationship) {
+	if (relationship == NULL)
+		return;
+
+	oscap_free(relationship->product_reference);
+	oscap_free(relationship->relates_to_ref);
+	cvrf_product_name_free(relationship->full_name);
+	oscap_free(relationship);
+}
+
+struct cvrf_relationship *cvrf_relationship_clone(const struct cvrf_relationship *relation) {
+	struct cvrf_relationship *clone = oscap_calloc(1, sizeof(struct cvrf_relationship));
+	clone->relation_type = relation->relation_type;
+	clone->product_reference = oscap_strdup(relation->product_reference);
+	clone->relates_to_ref = oscap_strdup(relation->relates_to_ref);
+	clone->full_name = cvrf_product_name_clone(relation->full_name);
+	return clone;
+}
+
+/***************************************************************************
+ * CVRF Branch
+ */
+struct cvrf_branch {
+	cvrf_branch_type_t branch_type;
+	char *branch_name;
+	struct cvrf_product_name *full_name;
+	struct oscap_list *subbranches;
+};
+OSCAP_ACCESSOR_STRING(cvrf_branch, branch_name)
+
+struct oscap_iterator *cvrf_branch_get_subbranches(struct cvrf_branch *branch) {
+	return oscap_iterator_new(branch->subbranches);
+}
+struct cvrf_product_name *cvrf_branch_get_cvrf_product_name(struct cvrf_branch *branch) {
+	return branch->full_name;
+}
+cvrf_branch_type_t cvrf_branch_get_branch_type(struct cvrf_branch *branch) {
+	return branch->branch_type;
+}
+
+struct cvrf_branch *cvrf_branch_new() {
+	struct cvrf_branch *ret = oscap_alloc(sizeof(struct cvrf_branch));
+	if (ret == NULL)
+		return NULL;
+
+	ret->branch_name = NULL;
+	ret->full_name = cvrf_product_name_new();
+	ret->subbranches = oscap_list_new();
+	return ret;
+}
+
+void cvrf_branch_free(struct cvrf_branch *branch) {
+	if (branch == NULL)
+		return;
+
+	oscap_free(branch->branch_name);
+	cvrf_product_name_free(branch->full_name);
+	oscap_list_free(branch->subbranches, (oscap_destruct_func) cvrf_branch_free);
+	oscap_free(branch);
+}
+
+struct cvrf_branch *cvrf_branch_clone(const struct cvrf_branch *branch) {
+	struct cvrf_branch *clone = oscap_calloc(1, sizeof(struct cvrf_branch));
+	clone->branch_name = oscap_strdup(branch->branch_name);
+	clone->branch_type = branch->branch_type;
+	clone->full_name = cvrf_product_name_clone(branch->full_name);
+	clone->subbranches = oscap_list_clone(branch->subbranches, (oscap_clone_func) cvrf_branch_clone);
+	return clone;
+}
+
+static const char *get_cvrf_product_id_from_branch(struct cvrf_branch *branch, const char *cpe) {
+	const char *product_id = NULL;
+	if (cvrf_branch_get_branch_type(branch) == CVRF_BRANCH_PRODUCT_FAMILY) {
+		struct oscap_iterator *subbranches = cvrf_branch_get_subbranches(branch);
+		while(oscap_iterator_has_more(subbranches) && product_id == NULL) {
+			product_id = get_cvrf_product_id_from_branch(oscap_iterator_next(subbranches), cpe);
```

Comment:
There is a one-to-one CPE-to-ProductId relationship for branches. So it will find the only ProductID that matches the CPE name, if there is one